### PR TITLE
Refactors using explicit lookup patterns

### PR DESCRIPTION
**Pull request by mrmr1993**
This PR builds upon #581 to simplify the logic around the 'lookup constraint system' using the explicit lookup patterns.

### CODE REVIEW

File: kimchi/src/circuits/lookup/constraints.rs [hunk 1]
```diff
@@ -78,6 +78,7 @@ pub fn zk_patch<R: Rng + ?Sized, F: FftField>(
 //~
 
 /// Computes the sorted lookup tables required by the lookup argument.
+#[allow(clippy::too_many_arguments)]
```
Comment (mimoo) [inline kimchi/src/circuits/lookup/constraints.rs:81]: 🥶

File: kimchi/src/circuits/lookup/lookups.rs [hunk 2]
```diff
@@ -340,6 +340,24 @@ impl LookupPattern {
             }
         }
     }
+
+    pub fn table(&self) -> Option<GateLookupTable> {
```
Comment (mimoo) [inline kimchi/src/circuits/lookup/lookups.rs:342]: doc missing, suggest "Returns the lookup table used by the gate"

File: kimchi/src/circuits/lookup/lookups.rs [hunk 3]
```diff
@@ -263,13 +261,30 @@ impl<F: Copy> JointLookup<SingleLookup<F>, LookupTableID> {
...
 impl LookupPattern {
+    pub fn max_lookups_per_row(&self) -> usize {
+        match self {
+            LookupPattern::ChaCha => 4,
+            LookupPattern::ChaChaFinal => 4,
+            LookupPattern::LookupGate => 3,
+        }
+    }
+
+    pub fn max_joint_size(&self) -> u32 {
```
Comment (mimoo) [inline kimchi/src/circuits/lookup/lookups.rs:263]: missing doc, suggest "the maximum number of columns used by the tables of this lookup pattern"

File: kimchi/src/circuits/lookup/lookups.rs [hunk 4]
```diff
@@ -263,13 +261,30 @@ impl<F: Copy> JointLookup<SingleLookup<F>, LookupTableID> {
...
 
+#[derive(Clone, Serialize, Deserialize)]
 pub enum LookupPattern {
     ChaCha,
     ChaChaFinal,
     LookupGate,
 }
 
 impl LookupPattern {
+    pub fn max_lookups_per_row(&self) -> usize {
```
Comment (mimoo) [inline kimchi/src/circuits/lookup/lookups.rs:255]: missing doc, suggesting: "Returns the maximum number of lookups used by this pattern on a single row"

File: kimchi/src/circuits/lookup/lookups.rs [hunk 5]
```diff
@@ -263,13 +261,30 @@ impl<F: Copy> JointLookup<SingleLookup<F>, LookupTableID> {
...
+        }
+    }
+
+    pub fn max_joint_size(&self) -> u32 {
+        match self {
+            LookupPattern::ChaCha => 3,
+            LookupPattern::ChaChaFinal => 3,
+            LookupPattern::LookupGate => 2,
+        }
+    }
```
Comment (mimoo) [inline kimchi/src/circuits/lookup/lookups.rs:269]: both of these functions have hardcoded values, which makes me uneasy

Comment (mrmr1993) [reply kimchi/src/circuits/lookup/lookups.rs:272]: I've added a test that checks these constants against the computed values.

**Review by mimoo — APPROVED** [review]
Nice job!

**Review by jspada — COMMENTED** [review]
lgtm!

**Pull request merged**